Sources/NIOHTTP2/HTTP2StreamChannel.swift

```diff
@@ -163,34 +163,66 @@ final class HTTP2StreamChannel<Message: HTTP2FramePayloadConvertible & HTTP2Fram
         // 4. Catching errors if they occur.
-        let f = self.parent!.getOption(ChannelOptions.autoRead).flatMap { autoRead -> EventLoopFuture<Void> in
+        self.getAutoReadFromParent().flatMap { autoRead in
             self.autoRead = autoRead
             // This initializer callback can only be invoked if we already have a stream ID.
             // So we force-unwrap here.
             return initializer?(self, self.streamID!) ?? self.eventLoop.makeSucceededFuture(())
         }.map {
-            // This force unwrap is safe as parent is assigned in the initializer, and never unassigned.
-            // If parent is not active, we expect to receive a channelActive later.
-            if self.parent!.isActive {
-                self.performActivation()
-            }
+            self.activate(promise: promise)
+        }.whenFailure { error in
+            self.configurationFailed(withError: error, promise: promise)
+        }
+    }
 
-            // We aren't using cascade here to avoid the allocations it causes.
-            promise?.succeed(self)
+    internal func configure(initializer: ((Channel) -> EventLoopFuture<Void>)?, userPromise promise: EventLoopPromise<Channel>?){
+        // We need to configure this channel. This involves doing four things:
+        // 1. Setting our autoRead state from the parent
+        // 2. Calling the initializer, if provided.
+        // 3. Activating when complete.
+        // 4. Catching errors if they occur.
+        self.getAutoReadFromParent().flatMap { autoRead in
+            self.autoRead = autoRead
+            return initializer?(self) ?? self.eventLoop.makeSucceededFuture(())
+        }.map {
+            self.activate(promise: promise)
+        }.whenFailure { error in
+            self.configurationFailed(withError: error, promise: promise)
         }
+    }
 
-        f.whenFailure { (error: Error) in
-            switch self.state {
-            case .idle, .localActive, .closed:
-                // The stream isn't open on the network, nothing to close.
-                self.errorEncountered(error: error)
-            case .remoteActive, .active, .closing, .closingNeverActivated:
-                // In all of these states the stream is still on the network and we may need to take action.
-                self.closedWhileOpen()
-            }
+    /// Gets the 'autoRead' option from the parent channel.
+    private func getAutoReadFromParent() -> EventLoopFuture<Bool> {
+        // This force unwrap is safe as parent is assigned in the initializer, and never unassigned.
+        // Note we also don't set the value here: the additional `map` causes an extra allocation
+        // when using a Swift 5.0 compiler.
+        return self.parent!.getOption(ChannelOptions.autoRead)
+    }
 
-            promise?.fail(error)
+    /// Activates the channel if the parent channel is active and succeeds the given `promise`.
+    private func activate(promise: EventLoopPromise<Channel>?) {
```

Comment:
Can we give this a more evocative name? There's already a `performActivation` function here. I'd like this to indicate that it's called after the initializer fires, as part of initialisation.

Reply:
Not overly pleased with the new name. Suggestions gladly welcomed.

Reply:
Honestly I think the new name is fine.